polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1782,6 +1903,88 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
     return {{assigned, TypeAttr()}};
   }
 
+  /// Construct a DenseElementAttr to represent a constant array made up from \p
+  /// components.
+  static FailureOr<DenseElementsAttr>
+  getConstantArray(ArrayRef<ArrayComponent> components,
+                   LLVM::LLVMArrayType &arrTy) {
+    SmallVector<Attribute> constantValues(arrTy.getNumElements());
+    for (const auto &p : components) {
+      // Retrieve constant value.
+      SmallVector<OpFoldResult> foldResults;
+      if (failed(p.value.getDefiningOp()->fold({}, foldResults)) ||
+          !foldResults.front().is<Attribute>())
+        return failure();
+
+      Attribute constantComponent = foldResults.front().get<Attribute>();
+      size_t offset = p.offset;
+
+      // Retrieve individual elements of the constant array and collect them.
+      auto result =
+          TypeSwitch<Attribute, LogicalResult>(constantComponent)
+              .Case<DenseElementsAttr>([&](auto &denseAttr) {
+                for (auto attr : denseAttr.template getValues<Attribute>()) {
+                  constantValues[offset++] = attr;
+                }
+                return success();
+              })
+              .Case<IntegerAttr, FloatAttr>([&](auto &attr) {
+                constantValues[offset] = attr;
+                return success();
+              })
+              .Default(failure());
+
+      if (failed(result))
+        return failure();
+    }
+    return DenseElementsAttr::get(
+        RankedTensorType::get(arrTy.getNumElements(), arrTy.getElementType()),
+        constantValues);
+  }
+
+  /// Detect if the \p assigned value is a partial component of a bigger array.
+  /// Returns:
+  ///  * failure() if it is not part of an array
+  ///  * false if it is part of a bigger array, but not all array components
+  ///  have been found so far
+  ///  * true if \p assigned was the last missing part of the bigger array and
+  ///  set_captured should be inserted now.
+  static FailureOr<bool>
+  detectPartialArrayComponent(size_t offset, Operation *captureOp,
+                              Value assigned, Type expected,
+                              ArrayCollector &collector) {
+    if (!isa<LLVM::LLVMArrayType>(expected))
+      return failure();
+
+    auto arrayTy = cast<LLVM::LLVMArrayType>(expected);
+    size_t expectedNumElements = arrayTy.getNumElements();
+    Type expectedElemTy = arrayTy.getElementType();
+
+    auto numElements =
+        TypeSwitch<Type, size_t>(assigned.getType())
+            .Case<LLVM::LLVMFixedVectorType, VectorType>([&](const auto &ty) {
+              return (ty.getElementType() == expectedElemTy)
+                         ? ty.getNumElements()
+                         : 0;
+            })
+            .Default(
+                [&](const Type &ty) { return (ty == expectedElemTy) ? 1 : 0; });
```

Comment:
We can use value semantics with types
```suggestion
    auto numElements =
        TypeSwitch<Type, size_t>(assigned.getType())
            .Case<LLVM::LLVMFixedVectorType, VectorType>([=](auto ty) {
              return (ty.getElementType() == expectedElemTy)
                         ? ty.getNumElements()
                         : 0;
            })
            .Default(
                [=](Type ty) { return (ty == expectedElemTy) ? 1 : 0; });
```